src/segment/writer/mod.rs

```diff
@@ -153,8 +164,24 @@ impl Writer {
             // of the same key
 
             if self.bloom_policy.is_active() {
-                self.bloom_hash_buffer
-                    .push(Builder::get_hash(&item.key.user_key));
+                if let Some(ref extractor) = self.prefix_extractor {
+                    // Add all prefixes to bloom filter
+                    // If extract returns empty iterator (out of domain), don't add to bloom filter
+                    let mut prefixes = extractor.extract(&item.key.user_key);
+                    if let Some(first) = prefixes.next() {
+                        // Key is in domain, add first prefix
+                        self.bloom_hash_buffer.push(Builder::get_hash(first));
+                        // Add remaining prefixes
+                        for prefix in prefixes {
+                            self.bloom_hash_buffer.push(Builder::get_hash(prefix));
+                        }
+                    }
+                    // If no prefixes (out of domain), don't add to bloom filter
```

Comment:
Why check for the first prefix separately? Instead of just:

```rs
for prefix in extractor.extract(&item.key.user_key) {
    self.bloom_hash_buffer.push(Builder::get_hash(prefix));
}
```